File: Client/packages/logger/checkCategoryEnabled.ts
import type { ApiSubcategory, LogCategory } from "./categories";
import {
  apiSubcategoryToConfigKey,
  categoryToConfigKey,
  isAlwaysEnabled,
  LOG_CATEGORIES,
} from "./categories";
import type { ApiSubcategoryConfig, LoggerConfig } from "./loggerTypes";

/**
 * Check if a category is enabled for the given config snapshot.
 * Handles defined categories, future JSON-only categories, and API subcategories.
 */
export function checkCategoryEnabled(
  config: LoggerConfig,
  category: LogCategory | string,
  subcategory?: ApiSubcategory
): boolean {
  if (!category || typeof category !== "string") {
    return false;
  }

  if (category in LOG_CATEGORIES) {
    const logCategory = category as LogCategory;
    if (isAlwaysEnabled(logCategory)) {
      return true;
    }
  }

  let configKey: string;
  try {
    if (category in LOG_CATEGORIES) {
      configKey = categoryToConfigKey(category as LogCategory);
    } else {
      configKey = category
        .toLowerCase()
        .split("_")
        .map((word, index) => (index === 0 ? word : word.charAt(0).toUpperCase() + word.slice(1)))
        .join("");
    }
  } catch {
    configKey = category
      .toLowerCase()
      .split("_")
      .map((word, index) => (index === 0 ? word : word.charAt(0).toUpperCase() + word.slice(1)))
      .join("");
  }

  const configValue = config[configKey as keyof LoggerConfig];

  if (category === "API" || configKey === "api") {
    if (typeof configValue === "object" && configValue !== null) {
      const apiConfig = configValue as ApiSubcategoryConfig;
      if (subcategory) {
        const subcategoryKey = apiSubcategoryToConfigKey(subcategory);
        return apiConfig[subcategoryKey as keyof ApiSubcategoryConfig] === true;
      }
      return (
        apiConfig.initialLoad === true ||
        apiConfig.polling === true ||
        apiConfig.pageMount === true ||
        apiConfig.other === true
      );
    }
    return configValue === true;
  }

  return configValue === true;
}

File: Server/logger/core/check_category_enabled.py
"""Category enablement checks for server logger."""

from ..config.config_model import LOG_LEVELS, LoggerConfig
from .categories import LogCategory, category_to_config_key, is_always_enabled


def is_log_level_enabled(config_level: str, message_level: str) -> bool:
    return LOG_LEVELS.get(message_level, 0) >= LOG_LEVELS.get(config_level, 0)


def check_category_enabled(config: LoggerConfig, category: LogCategory) -> bool:
    if is_always_enabled(category):
        return True
    config_key = category_to_config_key(category)
    return bool(getattr(config, config_key, False))


def should_emit_log(config: LoggerConfig, level: str, category: LogCategory) -> bool:
    if not check_category_enabled(config, category):
        return False
    return is_log_level_enabled(config.logLevel, level)
